File: problem/Luogu/B3718__1.cpp
//  B3718 扔骰子问题 1
//  16:30
#include <bits/stdc++.h>

using int64 = int64_t;
using i32 = int64_t;
constexpr int mod = 998244353;
constexpr char endl = '\n';
constexpr char enld = '\n';
constexpr int64 inf_i64 = 0x3f3f3f3f3f3f3f3f;
constexpr std::string OJ_NAME{"Luogu"};

constexpr int P = 998244353;
using i64 = long long;
// assume -P <= x < 2P
int norm(int x) {
    if (x < 0) {
        x += P;
    }
    if (x >= P) {
        x -= P;
    }
    return x;
}

template<class T>
T power(T a, i64 b) {
    T res = 1;
    for (; b; b /= 2, a *= a) {
        if (b % 2) {
            res *= a;
        }
    }
    return res;
}

struct Z {
    int x;

    Z(int x = 0) : x(norm(x)) {
    }

    Z(i64 x) : x(norm(x % P)) {
    }

    int val() const {
        return x;
    }

    Z operator-() const {
        return Z(norm(P - x));
    }

    Z inv() const {
        assert(x != 0);
        return power(*this, P - 2);
    }

    Z &operator*=(const Z &rhs) {
        x = i64(x) * rhs.x % P;
        return *this;
    }

    Z &operator+=(const Z &rhs) {
        x = norm(x + rhs.x);
        return *this;
    }

    Z &operator-=(const Z &rhs) {
        x = norm(x - rhs.x);
        return *this;
    }

    Z &operator/=(const Z &rhs) {
        return *this *= rhs.inv();
    }

    friend Z operator*(const Z &lhs, const Z &rhs) {
        Z res = lhs;
        res *= rhs;
        return res;
    }

    friend Z operator+(const Z &lhs, const Z &rhs) {
        Z res = lhs;
        res += rhs;
        return res;
    }

    friend Z operator-(const Z &lhs, const Z &rhs) {
        Z res = lhs;
        res -= rhs;
        return res;
    }

    friend Z operator/(const Z &lhs, const Z &rhs) {
        Z res = lhs;
        res /= rhs;
        return res;
    }

    friend std::istream &operator>>(std::istream &is, Z &a) {
        i64 v;
        is >> v;
        a = Z(v);
        return is;
    }

    friend std::ostream &operator<<(std::ostream &os, const Z &a) {
        return os << a.val();
    }
};

struct Comb {
    int n;
    std::vector<Z> _fac;
    std::vector<Z> _invfac;
    std::vector<Z> _inv;

    Comb() : n{0}, _fac{1}, _invfac{1}, _inv{0} {
    }

    Comb(int n) : Comb() {
        init(n);
    }

    void init(int m) {
        if (m <= n) return;
        _fac.resize(m + 1);
        _invfac.resize(m + 1);
        _inv.resize(m + 1);

        for (int i = n + 1; i <= m; i++) {
            _fac[i] = _fac[i - 1] * i;
        }
        _invfac[m] = _fac[m].inv();
        for (int i = m; i > n; i--) {
            _invfac[i - 1] = _invfac[i] * i;
            _inv[i] = _invfac[i] * _fac[i - 1];
        }
        n = m;
    }

    Z fac(int m) {
        if (m > n) init(2 * m);
        return _fac[m];
    }

    Z invfac(int m) {
        if (m > n) init(2 * m);
        return _invfac[m];
    }

    Z inv(int m) {
        if (m > n) init(2 * m);
        return _inv[m];
    }

    Z binom(int n, int m) {
        if (n < m || m < 0) return 0;
        return fac(n) * invfac(m) * invfac(n - m);
    }
} comb(5e6 + 10);

i32 ans = 0;

std::vector<int> pow5(5e6 + 10);
std::vector<int> inv6(5e6 + 10);

signed main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    int T = 1;
    if (OJ_NAME != "AtCoder")
        std::cin >> T;
    pow5[0] = 1;
    for (int i = 1; i < 5e6 + 10; i++) {
        pow5[i] = 1ll * pow5[i - 1] * 5 % mod;
    }
    inv6[5e6 + 9] = power(Z(166374059), 5e6 + 9).val();
    for (int i = 5e6 + 8; i >= 0; i--) {
        inv6[i] = 1ll * inv6[i + 1] * 6 % mod;
    }

    while (T--) {
        int n, m;
        std::cin >> n >> m;
        ans ^= (comb.binom(n, m) * inv6[n] * pow5[n - m]).val();
    }
    std::cout << "tou\n" << ans << endl;
    return 0;
}

File: problem/Luogu/B3718__1.rs
//  B3718 扔骰子问题 1

use std::io::{self, Read};

struct Scanner {
    tokens: std::vec::IntoIter<String>,
}

impl Scanner {
    fn new() -> Self {
        let mut content = String::new();
        io::stdin().read_to_string(&mut content).unwrap();
        let tokens: Vec<String> = content
            .split_whitespace()
            .map(|s| s.to_owned())
            .collect();
        Self { tokens: tokens.into_iter() }
    }

    fn next<T: std::str::FromStr>(&mut self) -> T {
        self.tokens.next().unwrap().parse().ok().expect("Parse error")
    }
}
const P: i32 = 998244353;
fn qpow(mut a: i32, mut b: i32) -> i32 {
    if P == 1 { return 0; }
    let mut res = 1;
    a %= P;

    while b > 0 {
        if b % 2 == 1 {
            res = (res as i64 * a as i64 % P as i64) as i32;
        }
        a = (a as i64 * a as i64 % P as i64) as i32;
        b /= 2;
    }
    res as i32
}
fn main() {
    const N: i32 = 5000010;
    let mut pow5 = vec![0, N];
    let mut inv6 = vec![0, N];
    for i in 1..N {
        pow5[i] = pow5[i - 1] as i64 * 5 % P;
    }
    let mut cin = Scanner::new();
    let t: i32 = cin.next();
    let mut ans = 0;
    for _ in 0..t {
        let n: i32 = cin.next();


        let m: i32 = cin.next();
        ans ^= {
            qpow(qpow(6, P - 2), m) as i64 * qpow(5 * qpow(6, P - 2), n - m) as i64 % P as i64
        }
    }
    println!("tou");
    println!("{}", ans);
}
